src/context/AppProvider/hooks/useDiscoverAccounts.tsx

```diff
@@ -0,0 +1,138 @@
+            return { accountMetadataByAccountId: {}, hasActivity: false }
+          }
+
+          const walletId = await wallet.getDeviceID()
+          const isMultiAccountWallet = wallet.supportsBip44Accounts()
+          const currentPortfolio = portfolio.selectors.selectPortfolio(store.getState())
+
+          let accountNumber = 0
+          let hasActivity = true
+          const chainAccountMetadata: AccountMetadataById = {}
+
+          while (hasActivity) {
+            if (accountNumber > 0) {
+              if (!isMultiAccountWallet) break
+              if (isMetaMaskMultichainWallet && !isSnapInstalled) break
+            }
+
+            try {
+              const accountIdWithActivityAndMetadata = await getAccountIdsWithActivityAndMetadata(
+                accountNumber,
+                chainId,
+                wallet,
+                Boolean(isSnapInstalled),
+              )
+
+              hasActivity = accountIdWithActivityAndMetadata.some(account => account.hasActivity)
+
+              if (hasActivity || accountNumber === 0) {
+                accountIdWithActivityAndMetadata.forEach(({ accountId, accountMetadata }) => {
+                  chainAccountMetadata[accountId] = accountMetadata
+                })
+              }
+
+              accountNumber++
+            } catch (error) {
+              console.error(`Error discovering accounts for chain ${chainId}:`, error)
+              break
+            }
+          }
+
+          if (Object.keys(chainAccountMetadata).length > 0) {
+            dispatch(
+              portfolio.actions.upsertAccountMetadata({
+                accountMetadataByAccountId: chainAccountMetadata,
+                walletId,
+              }),
+            )
+
+            Object.keys(chainAccountMetadata).forEach(accountId => {
+              // Don't enable accounts that are already in the portfolio so we keep it disabled if user manually disabled it
+              if (
+                currentPortfolio.accountMetadata.byId[accountId] &&
+                currentPortfolio.wallet.byId[walletId]?.includes(accountId)
+              )
+                return
+
+              dispatch(portfolio.actions.enableAccountId(accountId))
+            })
+          }
+
+          return { accountMetadataByAccountId: chainAccountMetadata, hasActivity }
+        },
+        staleTime: Infinity,
+        gcTime: Infinity,
+        enabled: Boolean(wallet && deviceId),
+      })),
+    [dispatch, isSnapInstalled, wallet, deviceId, supportedChainIds, connectedRdns],
+  )
+
+  const accountsDiscoveryQueries = useQueries({
+    queries,
+  })
+
+  const { isLoading, isFetching } = useMemo(() => {
+    return accountsDiscoveryQueries.reduce(
+      (acc, query) => {
+        acc.isLoading = acc.isLoading || query.isLoading
+        acc.isFetching = acc.isFetching || query.isFetching
+        return acc
+      },
+      { isLoading: false, isFetching: false },
+    )
+  }, [accountsDiscoveryQueries])
```

Comment:
Also totally forgot to send that one but, looks like `.some()` would be more elegant if we want to do that?
Sure, that'd be *2 iterations but O(n) *2 but n is so smol anyway and so is the space complexity, literally just an accessooor (and akschually, not that it matters but hey I like trains: could be more performant, as `Array.prototype.some()` will short-circuit at the first occurence)